### ISSUE TITLE
Keras 3 Attention layer value tensor dimension

### ISSUE THREAD

**Issue by kwchan7** [keras-team/keras#19257]
hi,

I found the  below would not return the proper size output in Keras 3 (but works fine in Keras 2)
Please help to fix it, 

Thanks.

```python
import keras
from keras import layers

i = layers.Input((8,4))
xq = layers.Conv1D(5,1)(i)
xk = layers.Conv1D(5,1)(i)
xv = layers.Conv1D(7,1)(i)
o = layers.Attention()([xq,xv,xk])

m = keras.Model(inputs=i, outputs=o)
m.summary()

Output as below
┏━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━┓
┃ Layer (type)        ┃ Output Shape      ┃    Param # ┃ Connected to      ┃
┡━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━┩
│ input_layer         │ (None, 8, 4)      │          0 │ -                 │
│ (InputLayer)        │                   │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ conv1d (Conv1D)     │ (None, 8, 5)      │         25 │ input_layer[0][0] │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ conv1d_2 (Conv1D)   │ (None, 8, 7)      │         35 │ input_layer[0][0] │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ conv1d_1 (Conv1D)   │ (None, 8, 5)      │         25 │ input_layer[0][0] │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ attention           │ **(None, 8, 5)**      │          0 │ conv1d[0][0],     │
│ (Attention)         │                   │            │ conv1d_2[0][0],   │
│                     │                   │            │ conv1d_1[0][0]    │
└─────────────────────┴───────────────────┴────────────┴───────────────────┘
 Total params: 85 (340.00 B)
 Trainable params: 85 (340.00 B)
 Non-trainable params: 0 (0.00 B)
```

The Attention layer output shape should be (None, 8, 7), since **xv** is from **Conv1D** with 7 kernels.

Thie same code gives correct output from keras 2

```python
Model: "model"
__________________________________________________________________________________________________
 Layer (type)                   Output Shape         Param #     Connected to                     
==================================================================================================
 input_1 (InputLayer)           [(None, 8, 4)]       0           []                               
                                                                                                  
 conv1d (Conv1D)                (None, 8, 5)         25          ['input_1[0][0]']                
                                                                                                  
 conv1d_2 (Conv1D)              (None, 8, 7)         35          ['input_1[0][0]']                
                                                                                                  
 conv1d_1 (Conv1D)              (None, 8, 5)         25          ['input_1[0][0]']                
                                                                                                  
 attention (Attention)          **(None, 8, 7)**         0           ['conv1d[0][0]',                 
                                                                  'conv1d_2[0][0]',               
                                                                  'conv1d_1[0][0]']               
                                                                                                  
==================================================================================================
Total params: 85
Trainable params: 85
Non-trainable params: 0
```

**Comment by sampathweb** [comment]
```
inp = np.random.random((1, 8, 4))
m.build(np.random.random((1, 8, 4)))
m.compile()
m.summary()
out = m.predict(inp)
print(out.shape)  # Prints the correct shape of (1, 8, 5) - which is the dimension of Value vector.
```
But when we do model.summary(), it get `compute_output_shape` and that uses the shape of query vector - https://github.com/keras-team/keras/blob/master/keras/layers/attention/attention.py#L245

I checked with `tf-keras`.  It has similar code, but is able to infer the right output shape while building - https://github.com/keras-team/tf-keras/blob/master/tf_keras/layers/attention/base_dense_attention.py#L212

Both `keras 3` and `tf-keras` gives the right output shape which is `(b, seq, value_dim)` and they both enforce that query and key dim match since its required to obtain scores. So the outputs are the same. But Keras 3, uses query dim in summary for attention layer, whereas `tf-keras` correctly infers it as value dim.

@fchollet - Any thoughts on why this behavior is happening? I could change `compute_output_shape` to use `input_shape[1] (value dim)`, but this doesn't seem to be necessary for `tf-keras`, so not sure why there's a difference in behavior.

**Comment by kwchan7** [comment]
I think the expected out.shape is (1,8,7)
since output from the attention layer should follow the shape of "value",  and since
`xv = layers.Conv1D(7,1)(i)`

**Comment by sampathweb** [comment]
Right, it was a typo in my comment.  The output shape is (1,8,7) if you use the code snippet above.  Its the model.summary that shows the incorrect output shape.

**Comment by kwchan7** [comment]
I see.  But when I add more layers after the attention layer, error shows up.
code

```
import tensorflow as tf
import keras
from keras import layers
import numpy as np

print(keras.__version__)

i = layers.Input((8,4))
xq = layers.Conv1D(5,1)(i)
xk = layers.Conv1D(5,1)(i)
xv = layers.Conv1D(7,1)(i)
o1 = layers.Attention()([xq,xv,xk])
o2 = layers.Flatten()(o1)
o = layers.Dense(1)(o2)

m = keras.Model(inputs=i, outputs=o)
m.summary()

inp = np.random.random((1, 8, 4))
out = m.predict(inp)
print(out.shape)
```

output

```
Model: "functional_1"
┏━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━┓
┃ Layer (type)        ┃ Output Shape      ┃    Param # ┃ Connected to      ┃
┡━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━┩
│ input_layer         │ (None, 8, 4)      │          0 │ -                 │
│ (InputLayer)        │                   │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ conv1d (Conv1D)     │ (None, 8, 5)      │         25 │ input_layer[0][0] │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ conv1d_2 (Conv1D)   │ (None, 8, 7)      │         35 │ input_layer[0][0] │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ conv1d_1 (Conv1D)   │ (None, 8, 5)      │         25 │ input_layer[0][0] │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ attention           │ (None, 8, 5)      │          0 │ conv1d[0][0],     │
│ (Attention)         │                   │            │ conv1d_2[0][0],   │
│                     │                   │            │ conv1d_1[0][0]    │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ flatten (Flatten)   │ (None, 40)        │          0 │ attention[0][0]   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ dense (Dense)       │ (None, 1)         │         41 │ flatten[0][0]     │
└─────────────────────┴───────────────────┴────────────┴───────────────────┘
 Total params: 126 (504.00 B)
 Trainable params: 126 (504.00 B)
 Non-trainable params: 0 (0.00 B)
...
Input 0 of layer "dense" is incompatible with the layer: expected axis -1 of input shape to have value 40, but received input with shape (1, 56)

Arguments received by Functional.call():
  • inputs=tf.Tensor(shape=(1, 8, 4), dtype=float32)
  • training=False
  • mask=None
```

Basically, during model build, the dense() layer expect input of shape(40,), but receive shape(56,) during execution.

**Comment by kwchan7** [comment]
Sorry I may miss it, I see it is now closed, is a fix provided?

**Comment by kwchan7** [comment]
Never mind, see a fix is provided in https://github.com/keras-team/keras/pull/19284
